src/audio/kpb.c

```diff
@@ -608,18 +618,47 @@ static int kpb_copy(struct comp_dev *dev)
  * \param[in] source pointer to the buffer source.
  *
  */
-static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source,
+static int kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source,
 			    size_t size)
 {
+	int ret = 0;
 	size_t size_to_copy = size;
 	size_t space_avail;
 	struct hb *buff = kpb->history_buffer;
 	void *read_ptr = source->r_ptr;
+	size_t timeout = platform_timer_get(platform_timer) +
+			 clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1);
+	enum kpb_state state_preserved = kpb->state;
+	struct dd *draining_data = &kpb->draining_task_data;
 
 	tracev_kpb("kpb_buffer_data()");
 
+	if (kpb->state != KPB_STATE_RUN && kpb->state != KPB_STATE_HOST_COPY)
+		return PPL_STATUS_PATH_STOP;
+	else
+		kpb->state = KPB_STATE_BUFFERING;
+
+	if (kpb->state == KPB_STATE_DRAINING)
+		draining_data->buffered_while_draining += size_to_copy;
+
 	/* Let's store audio stream data in internal history buffer */
 	while (size_to_copy) {
+		/* Reset was requested, it's time to stop buffering and finish
+		 * KPB reset.
+		 */
+		if (kpb->state == KPB_STATE_RESETTING) {
+			kpb->state = KPB_STATE_RESET_FINISH;
+			kpb_reset(kpb->dev);
+			return PPL_STATUS_PATH_STOP;
+		}
+
+		/* Are we stuck in buffering? */
+		if (timeout < platform_timer_get(platform_timer)) {
+			trace_kpb_error("kpb_buffer_data(): stuck in "
```

Comment:
Maybe better just write *timeout* instead of *stuck in*, it's unusual to describe it like this in errmsg.

Reply:
Yes I agree, timeout will be better.